Automated Workflow for Agent-based Installer (#1067)
openshift/enhancements: enhancements/agent-installer/automated-workflow.md

- On-premises deployments, especially on bare metal, may need to take
  extra steps to fully utilize storage inside each host or visible to
  a host from a SAN. We therefore need to support configuration of
  per-host storage, including fibre channel settings such as
  multipath.
- We want to retain the OpenShift installer's ability to customize a
  deployment with arbitrary manifests.

### Non-Goals

- A user-driven workflow will be described in a separate enhancement.
- A user interface for collecting complex inputs such as per-host
  network settings before running the installation automatically is
  left to a future enhancement.
- Changes to the release image mirroring process are out of scope for
  this enhancement.
- Eliminating the requirement for a separate image registry to host
  the mirrored OpenShift image is left to a future enhancement.
- Describing a mechanism to add and remove nodes from the cluster after
  it is installed is left to a future enhancement.
- Generating image formats other than ISO is left to a future
  enhancement.
- Booting hosts via PXE is left to a future enhancement.
- Creating minimal ISO images with a separate rootfs is left to a

>   future enhancement.

Reviewer: I'm not sure this would require extra effort but we can leave it aside for now.

  assisted installer SaaS.
- Describing an end-to-end solution for users without their own
  orchestration tool is left to a future enhancement.
- Collecting telemetry data for connected deployments is left to a
  future enhancement.
- Details of securing the communication between agents and services to
  ensure only desired hosts join the cluster are left to a future
  enhancement.
- Details of a workflow for deploying in settings with DHCP assignment
  of IP addresses is left to a future enhancement.

## Proposal

We propose to extend the `openshift-install` binary with a new
sub-command, `create image` for generating a bootable image containing
all of the information needed to deploy a cluster.  A command line
tool avoids issues with prerequisites for deploying a long-running
service, and can be run on a low-resource host such as a laptop or
even on a host at a different site before deployment begins.

Most enterprise grade hardware includes a management controller that
can present an ISO image to a host virtually, allowing the host to
boot from the image remotely. Hypervisors such as kvm and vSphere also
support booting from ISOs. Therefore, we will concentrate on that

> format and approach first. Building minimal ISOs with separate root

Reviewer: Minimal ISO is very important for virtual media over slower networks but again, we can leave that out of scope for now.

filesystems or images that can be used with hosts that PXE boot will
come later.

We propose to leave up to the user's orchestration tool the
responsibilities of managing and storing the images, serving them on
the network, and especially configuring hosts to boot them. This gives
users the flexibility to perform those operations in any way they
like, with whatever tool they prefer. A design for an all encompassing
approach based on tools provided entirely by Red Hat is left for a
later enhancement.

> The `create image` command will need to take inputs similar to what

Author: @zaneb @celebdor you both had comments/concerns about the details for this aspect of the design. We may want to have a higher bandwidth conversation to work through those.

> The `create image` command will need to take inputs similar to what

Reviewer 2: what about multi version support? If it's part of the installer, how will the user install another version?

Author: I've added a non-goal for multi-version support at the end of the list (line 124 of the latest draft).

the baremetal IPI installer or ZTP APIs use to describe the cluster
(name, API IPs, etc.) and host-specific network settings (static IPs,

> VLANs, etc.). Using the existing file and custom resource types allows

Reviewer: That is true for the interactive flow, but in the automated flow, all inputs are provided upfront.  The main benefit of the agents in the interactive flow is validating as many as those inputs earlier on to fail quickly and with clear error messages.

> VLANs, etc.). Using the existing file and custom resource types allows

Reviewer 3: The agent pattern requires a live ISO to boot, pull container images (that contain the actual agent), and then for that agent to connect to the coordinating central service over the network. After that, validations and other configuration can happen. So the ISO must contain at least the minimum config for viable network connectivity.

> VLANs, etc.). Using the existing file and custom resource types allows

Author: I added the reason for needing the network settings.

us to use the same version management techniques that we do for the
APIs within a cluster.

> The `create image` command will produce a single image as output, to

Reviewer: I'm assuming that:
1. this is an RHCOS live ISO with embedded ignition
2. this is using coreos-installer to build the ISO and not Assisted?  The ignition should be compatible with Assisted though in terms of bringing up the agent, static networking, etc.

Reviewer 2: I assume we need some validation on available storage space on the laptop?

save storage space and reduce management complexity. The assisted
installer service can already generate images with boot scripts that
recognize which host is booting and apply host-specific network
settings and roles (such as running different services) accordingly,
and we will reuse that technique (and code).

One host, designated as "node0", will be configured to run the

> assisted service and a client that uses the assisted service API to

Reviewer: ```suggestion
assisted service and a client that uses the assisted service REST API to
```

automate the deployment.

> All other hosts will run the assisted agent, configured to talk to the

Reviewer: Clarify that node0 will also run the agent

service on node0.

> When all of the images are booted, the client on node0 will use the

Reviewer: ```suggestion
When all of the hosts are booted with the image, the client on node0 will use the
```

> When all of the images are booted, the client on node0 will use the

Reviewer: I'm assuming that (1) the input file lists the hosts and (2) the client is polling the inventory via REST API until all agents check in?

assisted service to trigger the deployment of the cluster (including
applicable pre-flight validations) without further interaction from
the user.

The `create image` command will produce as output credentials to

> access the cluster that is about to be created. This allows a client,

Reviewer 4: This is the advice from the docs.
```
It is recommended that you use Ignition config files within 12 hours after they are generated because the 24-hour certificate rotates from 16 to 22 hours after the cluster is installed. By using the Ignition config files within 12 hours, you can avoid installation failure if the certificate update runs during installation.
```

> access the cluster that is about to be created. This allows a client,

Reviewer 5: the final certs are generated by assisted-service calling openshift-installer in node0. In this stage, only the admin key and certificate should be the final ones. So, while more will be generated, the main task should be to find a way to set the CA generated in the create image step to be the same one that will be used to regenerate certs in the node0 stage.

such as a tool monitoring deployment progress, to use the cluster's
API as soon as the API service is available.

During deployment, the third-party orchestration tool will need a way
to monitor progress and discover errors. The assisted service
read-only API endpoints running on node0 will report that data. In the
event of failure, a read-only endpoint in that service will be
available to download any logs that have been collected.

We will be taking advantage of the assisted service's ability to run
the bootstrap services "in place" to avoid having a separate bootstrap
VM. Bootstrap will run on node0, where the assisted service runs, to
ensure that the service and its API are available for as long as
possible. When bootstrapping completes and node0 reboots, progress
monitoring will need to shift to the OpenShift APIs (`ClusterVersion`
and `ClusterOperator`). We will extend the `openshift-install wait-for

> install-complete` command to recognize when the cluster being deployed

Reviewer 6: I'm not sure that anything needs to be extended here. I think it would just work.

> install-complete` command to recognize when the cluster being deployed

Reviewer: I'm also a bit lost here.  Once node0 reboots, there are no Assisted components running anywhere (agents or service).  That means the logs/progress API endpoints described in the previous paragraph are no longer available.  In the paradigm where the service runs externally to the cluster, we do have a poorly-named component ("assisted controller", which is not a controller) that runs on the installed cluster and reports back to the service on progress and logs.
Now, in case of installation failure in the final steps, the user would need to understand that those REST endpoints disappear and they now need to use k8s APIs for further monitoring.  We need to be conscious of trading off this UX degradation with running the Assisted service externally to the cluster (on a laptop or single host that is used for all installations for the site).  We also need to remember that the Assisted Service currently provides credentials to the installed cluster, which are generated when the user clicks "install".

> install-complete` command to recognize when the cluster being deployed

Reviewer 2: Also, the users will have no access to logs post pivot.

> install-complete` command to recognize when the cluster being deployed

Reviewer 3: Just to be clear, the poorly-named component in the new cluster is a Job that completes eventually and goes away. So it's not a long-running component in the new cluster.

I think the point here is that it will get confused if it can't connect to the service, which will be the case when node0 reboots and starts joining the cluster. But its role in reporting on progress and errors prior to that pivot is probably still valuable (or maybe even required), so some work will be needed to adjust assumptions in one place or another to fit this new workflow.

> install-complete` command to recognize when the cluster being deployed

Reviewer 2: https://github.com/openshift/enhancements/pull/1067/files#r832051897

> install-complete` command to recognize when the cluster being deployed

Reviewer: Requiring people to use openshift-install to do any monitoring or debugging is a serious usability drawback.

> install-complete` command to recognize when the cluster being deployed

Reviewer 7: > Requiring people to use openshift-install to do any monitoring or debugging is a serious usability drawback.

I feel like this is a response to my question, but I do not understand it or how it applies to my question. Can you elaborate?

> install-complete` command to recognize when the cluster being deployed

Author: I've replaced `wait-for install-complete` with `agent wait-for`.

@avishayt , the user is not required to use the CLI to monitor progress. They can still use the REST API on node0 themselves, if that's what they prefer.

uses the agent installer and to use both APIs for monitoring and
[progress
reporting](https://github.com/openshift/assisted-service/blob/master/docs/enhancements/installation-progress-bar.md)
in that case, including reporting when steps like validation fail and
the cluster deployment is never going to succeed. Users will still be
able to write their own monitoring logic using the assisted service
and OpenShift APIs directly, but they will have the option of using
the command line tool.

> For the fully automated workflow, the steps a user might normally

Reviewer 2: another gap here is how the assisted-controller (bad name, it's just a job that finally completes) behaves.
Currently it approves nodes (in case of none platfrom, or always, I dont recall, send logs, progress reporting....

Reviewer 8: From my observation, the assisted-controller is deployed on the 2-node control-plane before the bootstrap node reboots to join the cluster. When the bootstrap node reboots, we lose assisted-service and assisted-controller is unable to update the service with progress or push logs to it. The cluster installation, however, continues to progress to completion. The assisted-controller continues to approve CSRs even when assisted-service can't be reached.

Author: I've clarified this statement to mention approving agents before installation starts, since that's what I was thinking of when I wrote it originally.

> perform such as approving hosts or triggering the installation will be

Reviewer 2: @staebler there is not create cluster stage in AI. only create manifests. And then those ignitions are propagated to the nodes

> perform such as approving hosts or triggering the installation will be

Reviewer 3: @staebler in the agent-based install pattern, the agents each phone home to the central service, send hardware details, do some validation, and then wait. They wait for the central service to tell them to start forming a cluster, and that service also tells each whether it will be the bootstrap host or not along with other critical info. The client described here is just performing that action of initiating the installation.

> perform such as approving hosts or triggering the installation will be

Reviewer 4: Perhaps we need to define more specific terms so that the term "installation" is not so overloaded.

> perform such as approving hosts or triggering the installation will be

Reviewer 5: >There's definitely some, uh, finessing to do to be able to use creds pre-generated and stored in the ISO before assisted runs openshift-install create manifests on the bootstrap smile

Yes. This is a very important gap that we must cover. To have the admin cred generated in ISO creation stage be valid after when assisted-service runs openshift-install again. This is currently far from existing.

handled by a new client running as a service on node0. The client will
read the user's inputs from files included in the image and interact
with the assisted-service REST API on localhost to drive the
deployment. For the purposes of this document, we will refer to that
service as a single entity called `create-cluster`, although in
practice we may divide some of the steps into separate tools.

### Actors in the Workflow

> **data center admin** is a human user responsible for infrastructure

Reviewer: We called this Infrastructure Admin in the past

Author: I've updated to use the same role names.

within the data center, such as an image registry, networking, etc.

> **cluster admin** is a human user responsible for deploying a cluster.

Reviewer: We called this Cluster Creator in the past

**orchestration tool** is intended to be a software component, but the

> steps may also be performed by **cluster admin** if no software

Reviewer: I guess some of the steps, like booting the hardware, might be done by the data-center/infrastructure admin

Author: Yes, in a "semi-automated" mode that might be the case.

> component is available.

Reviewer 6: Do we need to clarify this isn't something we're providing?

Author: Reworded

**installer** is the OpenShift installer binary, `openshift-install`.

**node0** is the first host to be added to a cluster.

### Steps for Deploying a Disconnected Cluster

> 1. The data center admin launches an image registry visible to the

Reviewer: Not sure if this is the data center admin or the cluster admin

   hosts that will form the cluster and to a host that has internet
   access so the release image can be downloaded.
2. The cluster admin obtains the `oc` command line tool and uses `oc`
   to copy the OpenShift release image into the local image registry
   in one of the usual ways with [`oc adm
   mirror`](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-installation-images.html)
   or [the `oc mirror`
   plugin](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-disconnected.html).
3. The cluster admin obtains the OpenShift installer in one of [the
   usual
   ways](https://docs.openshift.com/container-platform/4.9/installing/installing_bare_metal/installing-bare-metal.html#installation-obtaining-installer_installing-bare-metal).
4. The cluster admin triggers their orchestration tool to run the
   installer.

> 5. The orchestration tool prepares an `install-config.yaml` and

Reviewer 4: I am leery of using install-config.yaml as the input to `create image`. I assume that many of the fields are irrelevant. Re-using the same file will put a burden on the user to understand which fields are meaningful in which context. It is already cumbersome to have some fields that are only meaningful in the IPI flow, where the documentation describing that they are ignored when doing a UPI install is lacking.

Reviewer 4: Let's take the "none" platform as an example. It is mentioned above that we want to provide the API VIP. That is not currently an option when using the "none" platform. So, would we create a field for the "none" platform where we do not have one now? Would we add a field to that new type for API VIP that would only apply to agent-based installs? How do you convey well to the user that the "none" platform field is only used when doing an agent-based install?

As for fields that would not apply to an agent-based install, let's look at the "baremetal" platform. The following seem like fields that are unlikely to be applicable--which are almost all of the fields in the baremetal field.
libvirtURI
clusterProvisioningIP
provisioningHostIP
bootstrapProvisioningIP
externalBridge
externalMACAddress
provisioningNetwork
provisioningBridge
provisioningMACAddress
provisioningNetworkInterface
provisioningNetworkCIDR
provisioningDHCPExternal
provisioningDHCPRange
bootstrapOSImage
clusterOSImage

Author: I think we agreed offline that we would use a separate file for fields needed only for the agent workflow and not for other workflows.

> 5. The orchestration tool prepares an `install-config.yaml` and

Reviewer: This is the only reference to install-config.yaml in the document.  What is it used for?  I thought the input would be the ZTP API?

7. The installer extracts a copy of the RHCOS ISO from the OpenShift
   release payload.
8. The installer combines the inputs provided by the user and produces
   an Ignition config.
9. The installer combines the Ignition config and the RHCOS ISO to
   create the bootable image configured to run the assisted service or
   agent based on the host that boots the image and writes the new
   image and cluster credentials to disk.
10. The orchestration tool copies the ISO to a location that the hosts
    or VMs can boot from it. For bare metal, that will generally be a
    web server visible to the management controllers in the bare metal
    hosts (because the management controllers are unlikely to be on
    the same network as the external interfaces for the cluster
    hosts). For vSphere or other hypervisors, that may be a filesystem
    on the hypervisor host.
11. The orchestration tool configures the host to boot the ISO. For
    bare metal, that means configuring the baseboard management
    controller (BMC) in each host to expose the ISO through the
    virtual media interface. For vSphere or other hypervisors, similar
    settings of the VM need to be adjusted.

> 12. The orchestration tool boots the host.

Author: ```suggestion
12. The orchestration tool boots the hosts.
```

13. The orchestration tool uses `openshift-install wait-for
    install-complete` to watch the deployment progress and wait for it
    to complete.
14. On each host, the startup script in the image selects the correct
    network settings based on the MAC addresses visible and applies
    the network settings for each interface.
15. On node0, the startup script in the image recognizes that it is on
    node0 and launches the assisted service and the create-cluster
    service to drive the deployment.
16. On node0, the create-cluster service waits for the assisted
    service to start, then creates an InfraEnv using the API.
17. On node0, the create-cluster service waits for all of the known
    agents to be registered with the service.
18. On all nodes (including node0), the startup script in the image
    connects to the assisted service on node0 to find the InfraEnv ID

>     for the cluster being deployed (there will only be 1). After it

Author: We're doing it this way now, but is that how we want to keep doing it? Could we pre-generate the ID and embed it in the image?

Reviewer 3: Pre-generating sounds reasonable.

Reviewer 8: The ID is randomly generated within assisted-service: https://github.com/openshift/assisted-service/blob/afd8849feee8c450ba02085f16e723429ae32a6f/internal/bminventory/inventory.go#L3985

To accept a pre-generated id, the /v2/infra-envs REST-API endpoint would need to be changed to allow the id to become one of the parameters. Is that possible?

Reviewer: Allowing users to specify the ID would be a security issue.  We use the InfraEnv ID as part of the agent authentication.  Deleting an InfraEnv is our way of revoking the agents' access.  Allowing (potentially malicious) users to create an InfraEnv with the same ID as a deleted one is dangerous.
@carbonin please keep me honest here

Reviewer 9: Yes, the InfraEnv ID is used in our authentication methods.

The agent also requires the InfraEnv ID (and auth token) as input and does not have permissions to list InfraEnvs thus cannot discover what InfraEnvs may be available in the environment.
I did notice this in the non-goals though:
> Details of securing the communication between agents and services to
  ensure only desired hosts join the cluster are left to a future
  enhancement.

What does that mean for your plans to support authentication/https between the agent and service? I see several ways that auth between the agent and service could complicate this procedure.

Reviewer 10: Given that we only intend to support a single InfraEnv in the ephemeral installer, we could pre-generate the ID and pass it in as e.g. an environment variable. It wouldn't have to become part of the user-accessible API.

    has the ID, the startup script launches the assisted agent with
    that ID and configured to connect to the assisted service on
    node0.
19. All of the agents register their hosts with the assisted service.

> 20. The assisted service performs its standard validations for all of the

Reviewer: This is similar to the ZTP flow.  It's automated, so we need all of the inputs up-front, but we need the agents to be up to perform anything beyond basic input validations.

> 20. The assisted service performs its standard validations for all of the

Author: I updated this step to give a few examples.

    hosts.
21. On node0, the create-cluster service sees that all known agents
    have registered and the hosts have passed validation. It uses the

>     assisted service API to trigger the deployment to proceed.

Author: @rwsu on an earlier draft you said

> Likely to be a different service, a initiate-cluster-install service. I don't know if you intended to go into this much detail, but at around the same time, something will select node0 to be the bootstrap node.

I can update this to refer to the separate service.

We should describe how node0 is selected as part of this set of steps. Is that worked out, yet?

Reviewer: I think this wasn't stated explicitly, but I assume the user specifies which host will be node0 in the configuration?  So that it's clear where the REST API endpoints will be?
Also, what would this look like with DHCP?

Author: The agent installer does need to be able to control which node is node0. It's not entirely clear that the user cares, although they might. I had thought that we would either ask for an IP (for the DHCP case) or use the first host in the inventory (for the static networking case). But, we haven't decided which approach makes the most sense.

Reviewer 5: > Also, what would this look like with DHCP?

There's the requirement that at least one node should have static addressing. If it's only one node, that one will be node0. There's no other discovery mechanism available.

Reviewer: That requirement is a difficult one and should be made explicit

Author: All nodes must have IPs that don't change, even in the DHCP case. That's a fundamental cube requirement -- nodes can't change their IP.

It's not required that any nodes have static networking provided to the installer for the DHCP case, we just need to know one of the IPs to use to pick a host to be node0. That IP can be assigned by the DHCP server.

> 22. The assisted service runs the OpenShift installer and performs the

Reviewer 11: I think this step actually talking about the agent running coreos-installer, so perhaps the wording can be clarified?

Author: I can update the wording, but I thought it ran the installer to get the manifests. Maybe it would be best to remove any reference to what it actually runs, since the point is that it just does what it does today.

    cluster deployment in the usual way, except that the nodes will
    use the ISOs they booted from to install RHCOS instead of fetching

>     a new one from the assisted service.

Author: ```suggestion
    a new image from the assisted service.
```

> 23. The `openshift-install wait-for install-complete` process sees the

Reviewer: ```suggestion
23. The `openshift-install wait-for-install-complete` process sees the
```

    OpenShift API become available and starts using it to watch the
    cluster complete deployment, combining the information provided by
    `ClusterVersion` and `ClusterOperator` resources with the assisted
    service progress API.
24. On node0, when all of the other hosts have been deployed and
    bootstrapping is complete, the agent reboots node0.

> 25. The `openshift-install wait-for install-complete` process loses

Reviewer: ```suggestion
25. The `openshift-install wait-for-install-complete` process loses
```

    the connection to the assisted service REST API and starts relying
    entirely on the OpenShift API for data.

> 26. The `openshift-install wait-for install-complete` process sees

Reviewer: ```suggestion
26. The `openshift-install wait-for-install-complete` process sees
```

    that the `ClusterVersion` API in the cluster shows that the

>     deployment has completed, and reports success then exits.

Reviewer: What if the installation fails at this point?  What if it failed before node0 rebooted?

Reviewer 5: if it's before node0 reboots, we'd get the info from assisted-service. If it's after that, there'll need to be a timeout condition seeing that the cluster api does not list node0 as a master (obviously with SNO we get nothing at that point).

27. The orchestration tool may take steps to clean up (removing the
    ISO, disconnecting it from the boot sequence for the VMs or BMCs,
    etc.).

TODO: Write the steps describing what happens when create-cluster sees

>       that a host has failed validation.

Reviewer 11: I think adding this section is important, since it's the main differentiator vs e.g adding bootstrap-in-place to the existing UPI flow.

Do we expect the validations to be exposed via the existing `wait-for` CLI, or something else?

Reviewer 5: wait-for (or a separate command) should have both assisted validation and progress awareness and report it clearly.

Author: I've added some error handling variations. I didn't repeat the full set of steps, to make the variations easier to understand.

### User Stories

- As an OpenShift infrastructure owner, I need to be able to integrate
  the installation of my first on-premises OpenShift cluster with my
  automation flows and tools.
- As an OpenShift infrastructure owner, I must be able to provide the
  definition of the cluster I want to deploy as input to the tool.
- As an OpenShift Infrastructure owner, I must be able to retrieve the
  validation errors from an in-progress or failed deployment in a

>   programmatic way.

Reviewer 11: I think defining what "in a programmatic way" means may be helpful (perhaps in the TODO about the failed-validation path above), e.g as an Infrastructure owner with some existing orchestration tool (ansible, etc) how will I automate this flow, and detect and respond to error conditions?

Reviewer 5: calling the openshift-install subcommands would be the easier, since they will know how to deal with both REST APIs. Alternatively, doing the monitoring directly using the REST APIs

Author: See the variations listed above starting around line 331.

different installation workflows, and want to avoid inventing a new
one. Therefore we will reuse existing inputs relevant to on-premises
deployment, borrowing from the IPI installer and other tools such as
the zero-touch provisioning (ZTP) system in multi-cluster engine
(MCE).

Users often do not want to store the credentials for the baseboard
management controllers (BMC) of their hosts inside of the cluster they
are deploying. Therefore we cannot rely on them to give us those
credentials during installation, and we must support tools outside of
OpenShift managing the hosts' boot process. This places more of the
burden of designing the end-to-end automated workflow on users or our
solution architects.

### Risks and Mitigations

The inputs to building an on-premises cluster, especially the per-host
static network settings, can be complex. Manually creating the inputs
to the installer provides many opportunities to introduce errors. For
the automated use case, we assume that users are likely to mitigate
this risk by also automating production of the input files using
templates or file generators. Other mitigation approaches, such as a

> GUI for constructing the files, are left to future enhancements.

Reviewer 11: I think giving some thought to static validation of network configuration would be worthwhile, I know this is potentially difficult, but as a user if I provided some bad config (duplicate IPs, mismatched vlans, etc) I'd much rather learn about that early in the process vs at runtime when it's harder to debug.

/cc @cybertron

Author: If we're already doing something like that for the IPI installer, we would reuse it. If not, then I think we should do it independently of this enhancement.

Reviewer 5: Zane's right. We are waiting to have the bindings by the end of the month.

Author: I added some details about static network validation. I think we're going to want a separate enhancement to describe how the use of tho Rust library will work.

  connected settings
- End user documentation, relative API stability
- Sufficient test coverage
- Gather feedback from users rather than just developers
- Enumerate service level indicators (SLIs), expose SLIs as metrics
- Write symptoms-based alerts for the component(s)

#### Tech Preview -> GA

- Must support deploying in fully disconnected settings
- Automated end-to-end tests
- More testing (upgrade, downgrade, scale)
- Sufficient time for feedback
- Available by default
- Backhaul SLI telemetry
- Document SLOs for the component
- Conduct load testing
- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

Does not apply for the installer.

### Version Skew Strategy

Does not apply for the installer.

### Operational Aspects of API Extensions

No API changes.

#### Failure Modes

No API changes.

#### Support Procedures

No API changes.

## Implementation History

A prototype of the `create image` command is available in the
`fleeting` repository:
https://github.com/openshift-agent-team/fleeting

## Drawbacks

N/A

> ## Alternatives

Reviewer: Run the assisted service outside the installed cluster

Author: That was covered extensively in the discussion on https://github.com/openshift/enhancements/blob/master/enhancements/agent-installer/agent-based-installer.md

see-also:
  - "/enhancements/agent-installer/agent-based-installer.md"
  - "/enhancements/installer/single-node-installation-bootstrap-in-place.md"
  - "/enhancements/network/baremetal-ipi-network-configuration.md"
  - "https://github.com/openshift/enhancements/pull/1034"
replaces: N/A
superseded-by: N/A
---

# Automated Workflow for Agent-based Installer

## Summary

"Cluster 0" deployments, for the first cluster in an environment, are
unique because they occur in situations where there may not be a lot
of other infrastructure to support long-running services normally
associated with automated provisioning for on-premises use
cases. Nevertheless, users and partners want to automate these
deployments. The assisted installer GUI provides an excellent user
experience for manually deploying clusters. This enhancement covers
the *automation* use case.

## Motivation

### Goals

- The assisted installer SaaS can be used to deploy in fully connected
  on-premises environments and its REST API can be used to automate
  those deployments. In this enhancement, we need to describe a
  workflow for automated deployment of OpenShift clusters in fully

>   disconnected settings.

Author: I think the plural is correct here, but I'll adjust the wording to avoid any confusion.

   or [the `oc mirror`
   plugin](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-disconnected.html).
3. The cluster creator obtains the OpenShift installer in one of [the
   usual
   ways](https://docs.openshift.com/container-platform/4.9/installing/installing_bare_metal/installing-bare-metal.html#installation-obtaining-installer_installing-bare-metal).
4. The cluster creator triggers their orchestration tool to run the
   installer.
5. The orchestration tool prepares an `install-config.yaml`, including
   `NMState` content to configure the static network settings for the
   hosts, using input data and templates.
6. The orchestration tool runs `openshift-install create image`.
7. The installer extracts a copy of the RHCOS ISO from the OpenShift
   release payload.
8. The installer combines the inputs provided by the user and produces
   an Ignition config.
9. The installer combines the Ignition config and the RHCOS ISO to
   create the bootable image configured to run the assisted service or

>    agent based on the host that boots the image and writes the new

Reviewer 3: All hosts should run the agent, and one of them additionally runs the assisted service.

Reviewer 5: That's right.

Author: I've clarified this with an and/or

    the same network as the external interfaces for the cluster
    hosts). For vSphere or other hypervisors, that may be a filesystem
    on the hypervisor host.
11. The orchestration tool configures the host to boot the ISO. For
    bare metal, that means configuring the baseboard management
    controller (BMC) in each host to expose the ISO through the
    virtual media interface. For vSphere or other hypervisors, similar
    settings of the VM need to be adjusted.
12. The orchestration tool boots the hosts.
13. The orchestration tool uses `openshift-install wait-for
    install-complete` to watch the deployment progress and wait for it
    to complete.
14. On each host, the startup script in the image selects the correct
    network settings based on the MAC addresses visible and applies
    the network settings for each interface.
15. On node0, the startup script in the image recognizes that it is on
    node0 and launches the assisted service and the create-cluster
    service to drive the deployment.
16. On node0, the create-cluster service waits for the assisted
    service to start, then creates an InfraEnv using the API.

> 17. On node0, the create-cluster service waits for all of the known

Reviewer 3: s/known/expected/ ? Also, how will it know which or how many agents to expect? Perhaps it just looks at the cluster definition.

Reviewer 5: it just goes by count from the cluster definition.

Reviewer 3: I suggest clarifying that here.

  this enhancement.
- Eliminating the requirement for a separate image registry to host
  the mirrored OpenShift image is left to a future enhancement.
- Describing a mechanism to add and remove nodes from the cluster after
  it is installed is left to a future enhancement.
- Generating image formats other than ISO is left to a future
  enhancement.
- Booting hosts via PXE is left to a future enhancement.
- Creating minimal ISO images with a separate rootfs is left to a
  future enhancement.
- We are not currently considering replacing the IPI installer or
  assisted installer SaaS.
- Describing an end-to-end solution for users without their own
  orchestration tool is left to a future enhancement.
- Collecting telemetry data for connected deployments is left to a
  future enhancement.
- Details of securing the communication between agents and services to
  ensure only desired hosts join the cluster are left to a future
  enhancement.
- Details of a workflow for deploying in settings with DHCP assignment
  of IP addresses is left to a future enhancement.

## Proposal

We propose to extend the `openshift-install` binary with a new

> sub-command, `create image` for generating a bootable image containing

Author: Good point. I've updated the doc to say `agent create-image` throughout instead.

**installer** is the OpenShift installer binary, `openshift-install`.

**node0** is the first host to be added to a cluster.

### Steps for Deploying a Disconnected Cluster

1. The infrastructure admin launches an image registry visible to the
   hosts that will form the cluster and to a host that has internet
   access so the release image can be downloaded.
2. The cluster creator obtains the `oc` command line tool and uses `oc`
   to copy the OpenShift release image into the local image registry
   in one of the usual ways with [`oc adm
   mirror`](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-installation-images.html)
   or [the `oc mirror`
   plugin](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-disconnected.html).
3. The cluster creator obtains the OpenShift installer in one of [the
   usual
   ways](https://docs.openshift.com/container-platform/4.9/installing/installing_bare_metal/installing-bare-metal.html#installation-obtaining-installer_installing-bare-metal).
4. The cluster creator triggers their orchestration tool to run the
   installer.

> 5. The orchestration tool prepares an `install-config.yaml`, including

Reviewer: I thought I wrote a comment about install-config yesterday, but it seems to have disappeared.  This is the only place in the document where install-config.yaml is mentioned, and it's not clear what it's for.  Is the input to this whole process going to be the install-config.yaml, and not ZTP manifests?  Are we going to turn install-config into ZTP manifests, which in turn get turned into install-config?

Reviewer 5: installConfig, to me, is optional in this step .If you provide it, it will be used to generate content for the ZTP manifests, which are then put into the ISO. If you provide ZTP manifests from the get go, no installconfig should be necessary, as it is later generated by assisted-service.

Reviewer: What's the point of having two APIs to the same tool?

Reviewer 2: AI generates install-config. What will happen if it will disagree with the provided one?
We discussed it in the past while providing hive apis and finally decided to allow only patches in order to avoid the bad ux and complexity of merging. Not sure I understand the reasoning to have it as the main input - it already lacks some of the functionality, so having ztp manifests is not something that we'll be able to avoid.

Reviewer 10: It's an open question (that's what https://github.com/openshift/enhancements/pull/1067#discussion_r831557200 is referring to).

> Are we going to turn install-config into ZTP manifests, which in turn get turned into install-config?

This is definitely an option we're considering.

Author: If we use install-config as the input, I don't really see the benefit of converting to ZTP APIs as an intermediate step. We've established that the asset graph management code in the installer doesn't let the user modify the manifests, only add to them. So, we could simply copy the user-provided install-config into the image, then read that to drive the REST API. The fact that the assisted-service is going to generate a different install-config based on those API calls so it can run the installer binary again is an implementation detail.

Reviewer 10: > We've established that the asset graph management code in the installer doesn't let the user modify the manifests, only add to them.

Did we establish that? I thought what we established was:

1. Both install-config and manifests present; install-config modified -> does the Right Thing
2. Both install-config and manifests present; _both_ modified -> manifest changes are overwritten(!)
3. Both install-config and manifests present; manifests modified -> modified manifests used(???)
4. Only manifests present: not possible today, but we can hack it so that the install-config is optional

If (3) doesn't work like that then I agree that would be a compelling reason to only use only the install-config and not have ZTP manifests in a create-manifests stage.

Author: What I heard @staebler say was that if you have an install-config then the installer will generate fresh copies of the other manifests in the dependency graph every time it runs. I see that `create manifests` removes the install-config input file, so it wouldn't be there the second time around, and so we could potentially do something similar. I'm still not sure I see the point, though. Do we expect the ZTP APIs to support features that aren't supported in the install-config? Why would we want to recommend to the user that they modify any intermediate files?

Reviewer 4: If ZTP manifest assets are dependencies of the install config asset, then the installer will consider the ZTP manifest assets dirty if the install-config.yaml file is present. As a consequence, the installer will re-generate the ZTP manifest assets, overwriting any changes made by the user to any asset files that the installer generates.

Reviewer: This was resolved, but I still didn't see any compelling reason to not use the ZTP API as originally discussed.

>    `NMState` content to configure the static network settings for the

Reviewer 5: assisted-service supports nmstateconfig for any platform. I don't immediately see why we would block that usage.

>    `NMState` content to configure the static network settings for the

Author: It may be confusing to a user to have 2 ways to provide the same network data when the platform setting is `baremetal`. Do we want to try to converge the 2 use cases? We could extend the IPI installer to also look for separate NMStateConfig resources when used in that mode, for example. Or have the agent installer optionally generate NMStateConfig manifests based on the install-config content.

>    `NMState` content to configure the static network settings for the

Author: Let's defer discussion of changes to the IPI installer to later.

   hosts, using input data and templates.
6. The orchestration tool runs `openshift-install create image`.

> 7. The installer extracts a copy of the RHCOS ISO from the OpenShift

Reviewer 7: Is the RHCOS ISO currently in the release payload?

Reviewer 3: Yes. That's a relatively new inclusion.

   release payload.
8. The installer combines the inputs provided by the user and produces
   an Ignition config.
9. The installer combines the Ignition config and the RHCOS ISO to
   create the bootable image configured to run the assisted service or
   agent based on the host that boots the image and writes the new
   image and cluster credentials to disk.
10. The orchestration tool copies the ISO to a location that the hosts
    or VMs can boot from it. For bare metal, that will generally be a
    web server visible to the management controllers in the bare metal
    hosts (because the management controllers are unlikely to be on
    the same network as the external interfaces for the cluster
    hosts). For vSphere or other hypervisors, that may be a filesystem
    on the hypervisor host.
11. The orchestration tool configures the host to boot the ISO. For
    bare metal, that means configuring the baseboard management
    controller (BMC) in each host to expose the ISO through the
    virtual media interface. For vSphere or other hypervisors, similar
    settings of the VM need to be adjusted.
12. The orchestration tool boots the hosts.

> 13. The orchestration tool uses `openshift-install wait-for

Reviewer 2: will it be modified to work with assisted reset api on the first stage and then move to kube?

Author: Yes, see the discussion around line 203-217 in the latest draft.

    install-complete` to watch the deployment progress and wait for it
    to complete.
14. On each host, the startup script in the image selects the correct
    network settings based on the MAC addresses visible and applies
    the network settings for each interface.
15. On node0, the startup script in the image recognizes that it is on
    node0 and launches the assisted service and the create-cluster
    service to drive the deployment.
16. On node0, the create-cluster service waits for the assisted
    service to start, then creates an InfraEnv using the API.
17. On node0, the create-cluster service waits for all of the known

>     agents to be registered with the service.

Reviewer: In this step (17) you're waiting for the agents to register, but the agents are started only in the next step (18)?

Author: I merged steps 17 and 18 into one and explained that they happen in parallel.

    has the ID, the startup script launches the assisted agent with
    that ID and configured to connect to the assisted service on
    node0.
19. All of the agents register their hosts with the assisted service.
20. The assisted service performs its standard validations (network
    connectivity, storage space, RAM, etc.) for all of the hosts.
21. On node0, the create-cluster service sees that all known agents
    have registered and the hosts have passed validation. It uses the
    assisted service API to trigger the deployment to proceed.
22. The assisted service performs the cluster deployment in its usual
    way, except that the nodes will use the ISOs they booted from to
    install RHCOS instead of fetching a new image from the assisted
    service.
23. The `openshift-install wait-for install-complete` process sees the
    OpenShift API become available and starts using it to watch the
    cluster complete deployment, combining the information provided by
    `ClusterVersion` and `ClusterOperator` resources with the assisted
    service progress API.
24. On node0, when all of the other hosts have been deployed and
    bootstrapping is complete, the agent reboots node0.

> 25. The `openshift-install wait-for install-complete` process loses

Reviewer 2: we need new sub command - wait-for-agent-bootstrap? :) or something similar?

Reviewer 2: also for showing events (what the progress will show? Events or cluster status?)
Also, another required sub command will be to download logs

Author: I think this is covered in the section describing the `agent wait-for` command, currently on lines 203-217.

to the installer provides many opportunities to introduce errors. For
the automated use case, we assume that users are likely to mitigate
this risk by also automating production of the input files using
templates or file generators. Other mitigation approaches, such as a
GUI for constructing the files, are left to future enhancements.

## Design Details

### Open Questions [optional]

1. What do we do with the installation logs?

   - Write them to disk on each host in a way that allows someone
     logging in to find them or allows a Pod to mount the directory
     they are in and `cat` them to be collected via the OpenShift API.
   - Expect the orchestration tool to download them before allowing
     deployment to continue.
   - Can we take advantage of or replicate the installer's [bootstrap
     service
     records](https://github.com/openshift/installer/blob/6d778f911e79afad8ba2ff4301eda5b5cf4d8e9e/docs/dev/bootstrap_services.md)
     behavior?

2. Is the workflow different for single-node? Do we need to start the
   service and go through the orchestration steps for one node? If we
   don't, how do we handle the validations?

> 3. How do we implement a feature flag in the installer?

Reviewer 7: We discussed using a build flag, which is something we already do for ARO & OKD.

      it has the ID, the startup script launches the assisted agent
      with that ID and configured to connect to the assisted service
      on node0.
18. All of the agents register their hosts with the assisted service.
19. The assisted service performs its standard validations (network
    connectivity, storage space, RAM, etc.) for all of the hosts.
20. On node0, the create-cluster service sees that all known agents
    have registered and the hosts have passed validation. It uses the
    assisted service API to trigger the deployment to proceed.
21. The assisted service performs the cluster deployment in its usual
    way, except that the nodes will use the ISOs they booted from to
    install RHCOS instead of fetching a new image from the assisted
    service.
22. The `openshift-install agent wait-for` process sees the OpenShift
    API become available and starts using it to watch the cluster
    complete deployment, combining the information provided by
    `ClusterVersion` and `ClusterOperator` resources with the assisted
    service progress API.
23. On node0, when all of the other hosts have been deployed and

>     bootstrapping is complete, the agent reboots node0.

Reviewer 12: After this reboot, how does the host that was node0 rejoin the cluster and in what form?

Author: node0 will always be a control plane node. I've added some details about that in lines 202-207 above and by adding a step below the reboot step.

> 24. The `openshift-install agent wait-for` process loses

Reviewer 12: Where is the `openshift-install agent wait-for` process expected to run? If it was mentioned earlier in the document, I missed it.

---
title: automated-workflow-for-agent-based-installer
authors:
  - "@dhellmann"
reviewers:
  - "@avishayt"
  - "@celebdor"
  - "@lranjbar"
  - "@mhrivnak"
  - "@patrickdillon"
  - "@pawanpinjarkar"
  - "@rwsu"
approvers:
  - "@zaneb"
api-approvers:
  - None
creation-date: 2022-03-17
last-updated: 2022-03-17
tracking-link:
  - https://issues.redhat.com/browse/AGENT-10
see-also:
  - "/enhancements/agent-installer/agent-based-installer.md"
  - "/enhancements/installer/single-node-installation-bootstrap-in-place.md"
  - "/enhancements/network/baremetal-ipi-network-configuration.md"
  - "https://github.com/openshift/enhancements/pull/1034"
  - "/enhancements/network/baremetal-ipi-network-configuration.md"
replaces: N/A
superseded-by: N/A
---

# Automated Workflow for Agent-based Installer

## Summary

"Cluster 0" deployments, for the first cluster in an environment, are
unique because they occur in situations where there may not be a lot
of other infrastructure to support long-running services normally
associated with automated provisioning for on-premises use
cases. Nevertheless, users and partners want to automate these

> deployments. The assisted installer GUI provides an excellent user
> experience for manually deploying clusters. This enhancement covers
> the *automation* use case.

Author: We expect to have another workflow with a UI for non-automated use, so this change isn't quite right. How about:
```suggestion
deployments. The assisted installer GUI provides an excellent user
experience for deploying connected clusters. This enhancement covers
the *automation* use case for disconnected clusters.
```

> the *automation* use case.

Reviewer: Sure

sub-command, `agent create-image`, for generating a bootable image
containing all of the information needed to deploy a cluster.  A
command line tool avoids issues with prerequisites for deploying a
long-running service, and can be run on a low-resource host such as a
laptop or even on a host at a different site before deployment begins.

Most enterprise grade hardware includes a management controller that
can present an ISO image to a host virtually, allowing the host to
boot from the image remotely. Hypervisors such as kvm and vSphere also
support booting from ISOs. Therefore, we will concentrate on that
format and approach first. Building minimal ISOs with separate root
filesystems or images that can be used with hosts that PXE boot will
come later.

We propose to leave up to the user's orchestration tool the
responsibilities of managing and storing the images, serving them on
the network, and especially configuring hosts to boot them. This gives
users the flexibility to perform those operations in any way they
like, with whatever tool they prefer. A design for an all encompassing
approach based on tools provided entirely by Red Hat is left for a
later enhancement.

> The `agent create-image` command will need to file-based inputs to

Reviewer: ```suggestion
The `agent create-image` command will need file-based inputs to
```

describe the cluster and node-specific configuration. Using the
existing file and custom resource types allows us to use the same
version management techniques that we do for the APIs within a
cluster.  We will use the `install-config.yaml` format for
cluster-wide settings. We will use a new file, tentatively named

> `agent-config.yaml`, for settings that apply to only agent-based

Reviewer: Doesn't this go against this stated goal?

> - We want to avoid inventing completely new ways to describe OpenShift
>   clusters. Therefore we will reuse existing inputs to the installer
>   and to tools such as the zero-touch provisioning (ZTP) system in
>   multi-cluster engine (MCE).

Author: The new file is an extension of install-config. It's similar to adding content directly to that file, but right now we're considering agent-specific settings out of scope for install-config itself.

save storage space and reduce management complexity. The assisted
installer service can already generate images with boot scripts that
recognize which host is booting and apply host-specific network
settings and roles (such as running different services) accordingly,
and we will reuse that technique (and code).

The image generated by the `agent create-image` command will be an RHCOS
live ISO with embedded Ignition config containing systemd units to
start the assisted service, agent, network, etc. running in containers
pulled from the OpenShift release payload. Using RHCOS ensures that
the same OS that will run OpenShift boots on the hosts. Using
containers in the release payload means that when the payload is
mirrored all of the tools needed to run the agent based deployment
workflows will be present in the mirror.  More details about the
various services follow below.

The image generated by the `agent create-image` command will also
contain the data in the input `install-config.yaml`,
`agent-config.yaml`, and other files. That data may be converted to
another format (we anticipate using the ZTP CRDs, since those map to

> the assisted-service API calls). The intermediate format will be

Reviewer: So install-config+agent-config+manifests+nmstate -> ZTP -> install-config+manifests

Why not just start with ZTP?

Author: The manifests are optional. Install-config combines data from InfraEnv, ClusterDeployment, and AgentClusterInstall, so we end up with fewer files overall. Having those things split out separately makes sense in the multi-cluster situation, where some of the data is shared between clusters, but this tool is doing 1 cluster at a time.

Reviewer: If the goal is fewer files, you can create a [SiteConfig](https://docs.openshift.com/container-platform/4.9/scalability_and_performance/ztp-deploying-disconnected.html#ztp-creating-the-siteconfig-custom-resources_ztp-deploying-disconnected) or simply define all of the CRs in a single yaml file.  This is defining a new API, and the reason is unclear.

considered another layer of control for the cluster creator, just as
the manifests written by the `create manifests` sub-command are. The
cluster creator will be able to generate those intermediate files with
the sub-command `agent create-manifests`.

> One host, designated as "node0", will be configured to run the

Reviewer: ```suggestion
One host, designated as "node0", will be configured to run the
```

Specify that it's designated by the user, and must be on a host with a known IP?

Author: Done in 5534d864

assisted service and a client that uses the assisted service REST API
to automate the deployment. Because we must support single-node and
compact deployments, the node selected as node0 must always be
configured as a control plane node. Selecting other control plane
nodes may be left to the user.

All hosts will run the assisted agent, configured to talk to the
service on node0.

When all of the hosts are booted with the image, the client on node0
will use the assisted service to trigger the deployment of the cluster
(including applicable pre-flight validations) without further
interaction from the user.

The `agent create-image` command will produce as output credentials to

> access the cluster that is about to be created. This allows a client,

Reviewer: Do we know how to get the credentials into the installed cluster in a supported way?

Author: We're still working on that.

to monitor progress and discover errors. The assisted service
read-only API endpoints running on node0 will report that data. In the
event of failure, a read-only endpoint in that service will be
available to download any logs that have been collected.

We will be taking advantage of the assisted service's ability to run
the bootstrap services "in place" to avoid having a separate bootstrap
VM. Bootstrap will run on node0, where the assisted service runs, to
ensure that the service and its API are available for as long as
possible.

When bootstrapping completes and node0 reboots, progress monitoring
will need to shift to the OpenShift APIs (`ClusterVersion` and
`ClusterOperator`). We will extend the `openshift-install` binary with
a new sub-command, `agent wait-for`, to use both APIs for monitoring
and [progress
reporting](https://github.com/openshift/assisted-service/blob/master/docs/enhancements/installation-progress-bar.md)
in that case, including reporting when steps like validation fail and
the cluster deployment is never going to succeed. *Users will still be
able to write their own monitoring logic using the assisted service

> and OpenShift APIs directly, but they will have the option of using

Reviewer: I think this will be so complicated that very few will do that (REST API, OpenShift API, switching from one to the other)

Author: I suspect you're right, that's why we're providing an implementation in the installer.

the command line tool.*

For the fully automated workflow, the steps a user might normally
perform such as approving agents be for installation or triggering the
installation will be handled by a new client running as a service on
node0. The client will read the user's inputs from files included in
the image and interact with the assisted-service REST API on localhost
to drive the deployment. For the purposes of this document, we will
refer to that service as a single entity called `create-cluster`,
although in practice we may divide some of the steps into separate
tools.

### Actors in the Workflow

**infrastructure admin** is a human user responsible for infrastructure
within the data center, such as an image registry, networking, etc.

**cluster creator** is a human user responsible for deploying a cluster.

**orchestration tool** is intended to be a software component provided
by the end user, but the steps may also be performed by **cluster

> creator** if no software component is available.

Author: I've updated the description of the orchestration tool in this paragraph via bb48e644. I don't want to clutter the workflow steps with lots of conditional statements about who could be doing what. They're written with the assumption that most of the process is automated, because that's what this doc is describing. In the manual workflow design doc we'll have a similar workflow with all human actors in each step.

### Steps for Deploying a Disconnected Cluster

1. The infrastructure admin launches an image registry visible to the
   hosts that will form the cluster and to a host that has internet
   access so the release image can be downloaded.
2. The cluster creator obtains the `oc` command line tool and uses `oc`
   to copy the OpenShift release image into the local image registry
   in one of the usual ways with [`oc adm
   mirror`](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-installation-images.html)
   or [the `oc mirror`
   plugin](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-disconnected.html).
3. The cluster creator obtains the OpenShift installer in one of [the
   usual
   ways](https://docs.openshift.com/container-platform/4.9/installing/installing_bare_metal/installing-bare-metal.html#installation-obtaining-installer_installing-bare-metal).
4. The cluster creator triggers their orchestration tool to run the
   installer.
5. The orchestration tool prepares an `install-config.yaml`, including
   `NMState` content to configure the static network settings for the

>    hosts, using input data and templates.

Author: How a third-party tool manages its data is out of scope for this doc. Maybe I should remove the details that are here and just say that the tool prepares the input files?

>    hosts, using input data and templates.

Author: Updated in bb48e644

    controller (BMC) in each host to expose the ISO through the
    virtual media interface. For vSphere or other hypervisors, similar
    settings of the VM need to be adjusted.
12. The orchestration tool boots the hosts.
13. The orchestration tool uses `openshift-install agent wait-for` to
    watch the deployment progress and wait for it to complete.
14. On each host, the startup script in the image selects the correct
    network settings based on the MAC addresses visible and applies
    the network settings for each interface.
15. On node0, the startup script in the image recognizes that it is on
    node0 and launches the assisted service and the create-cluster
    service to drive the deployment.
16. On node0, the create-cluster service waits for the assisted
    service to start, then creates an InfraEnv using the API.
17. In parallel
    - On node0, the create-cluster service waits for the number of
      agents registered with the service to match the expected value
      based on the input from the user.
    - On all nodes (including node0), the startup script in the image
      connects to the assisted service on node0 to find the InfraEnv

>       ID for the cluster being deployed (there will only be 1). After

Reviewer: We said that the InfraEnv ID will be baked into the ISO and also passed to the assisted-service, because once auth is enabled agents won't be able to list InfraEnvs.

Author: Yep, thanks, I missed updating this step.

Author: I updated this step and made a few related changes in 6ce0fe66

  Hypershift, which may be used for multi-cluster provisioning or
  management in the same environment.
- We want to minimize the amount of pre-deployment work that an admin
  has to do before they can start taking steps to deploy their
  cluster.
- Many of our users and partners want to own the overall orchestration
  for deployment, of which the OpenShift installer is only a small
  part of a larger process. To support that goal, we want to be
  agnostic to the tools used to provision (virtual) machines so that
  users can use their own (possibly pre-existing) tooling for
  provisioning.  This lets us support the widest variety of hardware
  and allows the user to work with tools that are already familiar to
  them. This approach also differentiates this workflow from the
  existing installer-provisioned infrastructure workflow, which has
  automation built into the installer.
- We want to minimize the extra hardware requirements for the process,
  by taking steps to eliminate the host needed to run a REST API
  service or act as a hypervisor for a bootstrap VM, for example.
- We want to install clusters with [single-node, compact, and
  high-availability

>   topologies](/enhancements/single-node/cluster-high-availability-mode-adpi.md).

Reviewer 13: Just `/enhancements/single-node` or `https://github.com/openshift/enhancements/tree/master/enhancements/single-node`

>   topologies](/enhancements/single-node/cluster-high-availability-mode-adpi.md).

Author: I want to link to the document that describes the topology API and what the topologies are, so it's not enough just to link to the `single-node` directory. I fixed the typo in 5d7aa027

the command line tool.*

For the fully automated workflow, the steps a user might normally
perform such as approving agents be for installation or triggering the
installation will be handled by a new client running as a service on
node0. The client will read the user's inputs from files included in
the image and interact with the assisted-service REST API on localhost
to drive the deployment. For the purposes of this document, we will
refer to that service as a single entity called `create-cluster`,
although in practice we may divide some of the steps into separate
tools.

### Actors in the Workflow

**infrastructure admin** is a human user responsible for infrastructure
within the data center, such as an image registry, networking, etc.

**cluster creator** is a human user responsible for deploying a cluster.

**orchestration tool** is intended to be a software component provided
by the end user for managing the overall deployment process, including
any steps for which the installer is not responsible. Those steps may
also be performed by **cluster creator** if no software component is
available.

**installer** is the OpenShift installer binary, `openshift-install`.

> **node0** is the first host to be added to a cluster.

Reviewer 13: More precisely, the node (identified by user) which runs assisted-service along with agent.

Reviewer 8: node0 is the last of the control plane nodes to join the cluster.

Author: I think here I meant it was the first in the list of hosts known to the installer, but I've rewritten the definition based on running the service using the same wording that appears earlier in this doc. 55e97611

1. The infrastructure admin launches an image registry visible to the
   hosts that will form the cluster and to a host that has internet
   access so the release image can be downloaded.
2. The cluster creator obtains the `oc` command line tool and uses `oc`
   to copy the OpenShift release image into the local image registry
   in one of the usual ways with [`oc adm
   mirror`](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-installation-images.html)
   or [the `oc mirror`
   plugin](https://docs.openshift.com/container-platform/4.10/installing/disconnected_install/installing-mirroring-disconnected.html).
3. The cluster creator obtains the OpenShift installer in one of [the
   usual
   ways](https://docs.openshift.com/container-platform/4.9/installing/installing_bare_metal/installing-bare-metal.html#installation-obtaining-installer_installing-bare-metal).
4. The cluster creator triggers their orchestration tool to run the
   installer.
5. The orchestration tool (or cluster creator) prepares an
   `install-config.yaml`, including `NMState` content to configure the
   static network settings for the hosts.

> 6. The orchestration tool runs `openshift-install create image`.

Reviewer 7: nit:
```suggestion
6. The orchestration tool runs `openshift-install agent create-image`.
```

   generated information like the InfraEnv ID and REST API credentials
   to produce an Ignition config.
9. The installer combines the Ignition config and the RHCOS ISO to
   create the bootable image configured to run the assisted service and/or
   agent based on the host that boots the image and writes the new
   image and cluster credentials to disk.
10. The orchestration tool copies the ISO to a location that the hosts
    or VMs can boot from it. For bare metal, that will generally be a
    web server visible to the management controllers in the bare metal
    hosts (because the management controllers are unlikely to be on
    the same network as the external interfaces for the cluster
    hosts). For vSphere or other hypervisors, that may be a filesystem
    on the hypervisor host.
11. The orchestration tool configures the host to boot the ISO. For
    bare metal, that means configuring the baseboard management
    controller (BMC) in each host to expose the ISO through the
    virtual media interface. For vSphere or other hypervisors, similar
    settings of the VM need to be adjusted.
12. The orchestration tool boots the hosts.

> 13. The orchestration tool uses `openshift-install agent wait-for` to

Reviewer 7: Perhaps `openshift-install agent wait-for-install-complete`

Author: I replaced `wait-for` with `wait-for-install-complete` in d88feb13.
